File: R/occlusion.R
#' What part of the record a fitted model reads
#'
#' Holds one bin of the record back at a time, rescores the held-out units, and records the fall in
#' score as that bin's weight. Nothing is refitted: the models kept by
#' `window_ladder(keep_fits = TRUE)` are the ones read, so the profile describes the models that
#' produced the reported scores rather than a fresh set of them.
#'
#' A model has to be shown something in place of a held-back bin, and what it is shown decides what
#' the weight means. Permuting the bin's values across units keeps the observed readings exactly
#' and cuts only the link between a reading and its unit. Replacing every unit by the fitting-fold
#' mean removes all between-unit variation while keeping the shape of the year. Replacing the bin
#' by each unit's own mean over the record keeps how warm a unit is and removes only that bin's
#' departure from it.
#'
#' Read with `over = "channel"` the same machinery asks what each statistic of a window carries,
#' holding one channel back across the whole record instead of one bin across all channels.
#'
#' @param ladder A [window_ladder()] result fitted with `keep_fits = TRUE`.
#' @param x The representation set the ladder was fitted on.
#' @param y The response it was fitted to.
#' @param arm The arm to read, as `"window|learner"` or `"learner"`.
#' @param over `"bin"` to hold each bin back in turn, `"channel"` for each channel.
#' @param substitute What a held-back part is replaced by: `"permute"`, `"fold_mean"` or
#'   `"unit_mean"`.
#' @param metric Name of the registered metric the rescoring is read by. The area under the ROC
#'   curve responds to every reordering of the units, where a maximum over thresholds frequently
#'   does not move at all, which is why it is the default here and not for the ladder.
#' @param permutations Draws averaged over, for `substitute = "permute"`.
#' @param seed Random seed.
#'
#' @return A data frame of one row per held-back part and variable, carrying the mean weight over
#'   folds and the score with and without the part.
#'
#' @examples
#' set.seed(1)
#' t <- seq(as.POSIXct("2021-09-01", tz = "UTC"), by = "hour", length.out = 24 * 120)
#' units <- sprintf("p%02d", 1:40)
#' warmth <- rnorm(40)
#' d <- data.frame(
#'   plot = rep(units, each = length(t)), t = rep(t, length(units)),
#'   temp = as.numeric(vapply(warmth, function(w) w + sin(seq_along(t) / 300) + rnorm(length(t)),
#'                            numeric(length(t)))))
#' y <- matrix(rbinom(80, 1, plogis(c(warmth, -warmth))), nrow = 40,
#'             dimnames = list(units, c("sp1", "sp2")))
#' x <- window_matrix(d, plot, t, temp, window = "month")
#' lad <- window_ladder(x, y, elasticnet_learner(), folds = fold_map(y, v = 3),
#'                      keep_fits = TRUE, verbose = FALSE)
#' head(bin_occlusion(lad, x, y, "month|elasticnet", permutations = 3))
#'
#' @export
bin_occlusion <- function(ladder, x, y, arm, over = c("bin", "channel"),
                          substitute = c("permute", "fold_mean", "unit_mean"),
                          metric = "roc_auc", permutations = 20L, seed = 1L) {
  over <- match.arg(over)
  substitute <- match.arg(substitute)
  fits <- attr(ladder, "fits")
  if (is.null(fits)) {
    stop("this ladder kept no fits. Refit with window_ladder(..., keep_fits = TRUE).",
         call. = FALSE)
  }
  rows <- .arm_rows(ladder, arm)
  label <- attr(rows, "label")
  window <- strsplit(label, "|", fixed = TRUE)[[1L]][1L]
  set <- .as_set(x)
  m <- set[[window]]
  if (is.null(m)) {
    stop("the representation carries no \"", window, "\" window.", call. = FALSE)
  }
  units <- dimnames(m)[[1L]]
  y <- .align_response(.responses_reg$get(attr(ladder, "response"))$prepare(y), units)
  f <- .as_folds(attr(ladder, "folds"), units)
  cells <- attr(ladder, "cells")
  score <- .metrics_reg$get(metric)

  parts <- if (over == "bin") seq_len(dim(m)[2L]) else seq_len(dim(m)[3L])
  labels <- if (over == "bin") dimnames(m)[[2L]] else dimnames(m)[[3L]]

  old <- .seed_state()
  on.exit(.restore_seed(old), add = TRUE)
  set.seed(seed)

  out <- list()
  for (k in sort(unique(f))) {
    fit <- fits[[paste(label, k, sep = "|")]]
    if (is.null(fit)) {
      next
    }
    test <- which(f == k)
    train <- which(f != k)
    base <- stats::predict(fit, .subset_units(m, test))
    ok <- .scorable_for(cells, colnames(y), k)
    full <- .score_columns(y[test, , drop = FALSE], base, ok, score)

    for (i in seq_along(parts)) {
      draws <- if (substitute == "permute") permutations else 1L
      acc <- matrix(0, nrow = draws, ncol = ncol(y))
      for (r in seq_len(draws)) {
        occluded <- .occlude(m, test, train, parts[i], over, substitute)
        acc[r, ] <- .score_columns(y[test, , drop = FALSE],
                                   stats::predict(fit, occluded), ok, score)
      }
      out[[length(out) + 1L]] <- data.frame(
        part = labels[i], variable = colnames(y), fold = k,
        score_full = full, score_held_back = colMeans(acc),
        weight = full - colMeans(acc), stringsAsFactors = FALSE)
    }
  }
  out <- do.call(rbind, out)
  out <- out[!is.na(out$weight), , drop = FALSE]
  agg <- stats::aggregate(out[c("score_full", "score_held_back", "weight")],
                          out[c("part", "variable")], mean)
  agg$part <- factor(agg$part, levels = labels)
  agg <- agg[order(agg$part, agg$variable, method = "radix"), ]
  agg$part <- as.character(agg$part)
  rownames(agg) <- NULL
  structure(agg, class = c("climgrain_occlusion", "data.frame"), arm = label, over = over,
            substitute = substitute, metric = metric)
}

.occlude <- function(m, test, train, i, over, substitute) {
  sub <- .subset_units(m, test)
  n <- dim(sub)[1L]
  b <- dim(sub)[2L]
  if (over == "channel") {
    sub[, , i] <- switch(
      substitute,
      permute = .plane(sub, i)[sample.int(n), , drop = FALSE],
      fold_mean = matrix(colMeans(.plane(m, i)[train, , drop = FALSE]),
                         nrow = n, ncol = b, byrow = TRUE),
      unit_mean = matrix(rowMeans(.plane(sub, i)), nrow = n, ncol = b)
    )
    return(sub)
  }
  for (ch in seq_len(dim(sub)[3L])) {
    sub[, i, ch] <- switch(
      substitute,
      permute = sub[sample.int(n), i, ch],
      fold_mean = mean(m[train, i, ch]),
      unit_mean = rowMeans(.plane(sub, ch))
    )
  }
  sub
}

# One channel of a representation as a [unit, bin] matrix, whatever the number of bins. Dropping to
# a vector at a single bin is what makes the row and column means below silently wrong.
.plane <- function(x, channel) {
  matrix(as.numeric(x[, , channel]), nrow = dim(x)[1L], ncol = dim(x)[2L])
}

.scorable_for <- function(cells, variables, fold) {
  cells$scorable[match(paste(variables, fold), paste(cells$variable, cells$fold))]
}

.score_columns <- function(y, p, ok, score) {
  vapply(seq_len(ncol(y)), function(j) {
    if (isTRUE(ok[j])) score(y[, j], p[, j]) else NA_real_
  }, numeric(1L))
}

#' @export
print.climgrain_occlusion <- function(x, ...) {
  cat("<climgrain occlusion>", attr(x, "arm"), "read by", attr(x, "metric"), "\n")
  cat("held back:", attr(x, "over"), "; substitute:", attr(x, "substitute"), "\n")
  w <- stats::aggregate(list(weight = x$weight), x["part"], mean)
  w <- w[order(-w$weight), ]
  cat("heaviest:", paste(utils::head(w$part, 5L), collapse = ", "), "\n")
  print(utils::head(as.data.frame(x), 6L))
  invisible(x)
}

File: python/timegrain/occlusion.py
"""What part of the record a fitted model reads, and what an already-reduced feature table is."""

from __future__ import annotations

from dataclasses import replace

import numpy as np

from .learners import Learner
from .registry import METRICS, get_learner
from .representation import WindowMatrix
from .response import Response, align_folds


def feature_matrix(m, units=None, features=None, label: str = "features") -> WindowMatrix:
    """Bring an already-reduced feature table into a ladder as a one-channel representation.

    It carries no time axis, because it has none: the reduction already happened, elsewhere, and
    what reaches the model is a list of numbers per unit. That is the whole point of comparing
    against it.
    """
    m = np.asarray(m, dtype=np.float64)
    n_u, n_f = m.shape
    units = tuple(units) if units is not None else tuple(str(i) for i in range(n_u))
    features = tuple(features) if features is not None else tuple(f"f{j}" for j in range(n_f))
    empty = np.full(n_f, np.datetime64("NaT"), dtype="datetime64[s]")
    return WindowMatrix(values=m.reshape(n_u, n_f, 1), units=units, bins=features,
                        stats=(label,), window=label, year_start="", bin_start=empty,
                        bin_end=empty, bin_n=np.zeros((n_u, n_f), dtype=np.int64),
                        bin_partial=np.zeros(n_f, dtype=bool))


def ensemble_learner(members, weights=None, name: str = "ensemble") -> Learner:
    """Average several learners' predicted probabilities before the threshold is chosen.

    The average is taken before any decision is made, so the set is scored as one model rather than
    as a vote between decisions. Choose the members on an inner validation split or on prior
    grounds, never on the held-out folds.
    """
    members = [get_learner(m) for m in members]
    if len(members) < 2:
        raise ValueError("an ensemble needs at least two members")
    w = np.ones(len(members)) if weights is None else np.asarray(weights, dtype=np.float64)
    if len(w) != len(members) or (w < 0).any() or w.sum() <= 0:
        raise ValueError("`weights` needs one non-negative number per member and cannot be all "
                         "zero")
    w = w / w.sum()

    def fit(x, y, **kwargs):
        return [m.fit(x, y, **{**m.params, **kwargs}) for m in members]

    def predict(model, x):
        return sum(w[k] * np.asarray(members[k].predict(model[k], x), dtype=np.float64)
                   for k in range(len(members)))

    needs = tuple({n for m in members for n in m.needs})
    return Learner(name=name, fit=fit, predict=predict, needs=needs)


def bin_occlusion(ladder, x, y: Response, arm: str, over: str = "bin",
                  substitute: str = "permute", metric: str = "roc_auc",
                  permutations: int = 20, seed: int = 1):
    """Hold one bin of the record back at a time and record the fall in score as its weight.

    Nothing is refitted: the models kept by ``window_ladder(keep_fits=True)`` are the ones read.
    What a held-back bin is replaced by decides what the weight means, so the substitute is part of
    the answer: ``permute`` keeps the observed readings and cuts only the link between a reading
    and its unit, ``fold_mean`` removes all between-unit variation, ``unit_mean`` keeps how warm a
    unit is and removes only that bin's departure from it.
    """
    if not ladder.fits:
        raise ValueError("this ladder kept no fits; refit with window_ladder(..., keep_fits=True)")
    if over not in ("bin", "channel"):
        raise ValueError(f"`over` must be 'bin' or 'channel', got {over!r}")
    if substitute not in ("permute", "fold_mean", "unit_mean"):
        raise ValueError(f"unknown substitute {substitute!r}")

    window, learner = arm.split("|", 1) if "|" in arm else (None, arm)
    label = f"{window}|{learner}"
    m = x if isinstance(x, WindowMatrix) else dict(x)[window]
    y = y.align(m.units).check_presence_absence()
    f = align_folds(ladder.folds, m.units)
    score = metric if callable(metric) else METRICS.get(metric)

    n_parts = m.values.shape[1] if over == "bin" else m.values.shape[2]
    labels = m.bins if over == "bin" else m.stats
    rng = np.random.default_rng(seed)

    weight = np.full((n_parts, len(y.variables)), np.nan)
    counted = np.zeros((n_parts, len(y.variables)))
    for k in np.unique(f):
        fit = ladder.fits.get(f"{label}|{k}")
        if fit is None:
            continue
        test = np.flatnonzero(f == k)
        train = np.flatnonzero(f != k)
        sub = m.take_units(test)
        base = np.asarray([score(y.values[test, j], fit.predict(sub)[:, j])
                           for j in range(len(y.variables))])
        for i in range(n_parts):
            draws = permutations if substitute == "permute" else 1
            acc = np.zeros((draws, len(y.variables)))
            for r in range(draws):
                occluded = _occlude(m, sub, train, i, over, substitute, rng)
                p = fit.predict(occluded)
                acc[r] = [score(y.values[test, j], p[:, j]) for j in range(len(y.variables))]
            fall = base - acc.mean(axis=0)
            ok = np.isfinite(fall)
            weight[i, ok] = np.where(np.isnan(weight[i, ok]), 0, weight[i, ok]) + fall[ok]
            counted[i, ok] += 1

    with np.errstate(invalid="ignore"):
        weight = weight / np.where(counted > 0, counted, np.nan)
    return {"part": list(labels), "variable": list(y.variables), "weight": weight}


def _occlude(m: WindowMatrix, sub: WindowMatrix, train, i, over, substitute, rng) -> WindowMatrix:
    values = sub.values.copy()
    n = values.shape[0]
    if over == "channel":
        if substitute == "permute":
            values[:, :, i] = values[rng.permutation(n), :, i]
        elif substitute == "fold_mean":
            values[:, :, i] = m.values[train, :, i].mean(axis=0)[None, :]
        else:
            values[:, :, i] = values[:, :, i].mean(axis=1)[:, None]
        return replace(sub, values=values)
    for ch in range(values.shape[2]):
        if substitute == "permute":
            values[:, i, ch] = values[rng.permutation(n), i, ch]
        elif substitute == "fold_mean":
            values[:, i, ch] = m.values[train, i, ch].mean()
        else:
            values[:, i, ch] = values[:, :, ch].mean(axis=1)
    return replace(sub, values=values)
